### ISSUE TITLE
Git-js Store

### ISSUE THREAD

**Issue by PseudoCorps** [techfort/LokiJS#121]
Enable Git style versioning to reduce network overhead on update 'pull'.  

This applies to use cases that may need to receive updates over the network. Loki has the ability to track and apply changes, which pretty much solves the 'push' aspect. However, to create the changes there is currently the necessity of a direct local comparison. This means at some point the client side has to either dump the data or send the entire set over.
I think it would be very useful to store version information that a server could use to dynamically diff the objects and return patches or changes that Loki can then apply. 

My thought on the possible solutions were -
Tweak the current Loki changes API to dynamically track App Level* version hashes
Or 
Create a JS Git store. https://github.com/creationix/js-git

(sorry, hope its okay to post potential enhancements)

**Comment by techfort** [comment]
Well this is incredible timing as I was just reading a mail about this. This interests me a lot, obviously it requires a very careful design, but yes, i'm thinking of pushing this to top priority.

**Comment by PseudoCorps** [comment]
Awesome
I agree, it seems like there would likely be some design decisions that could:
broaden the scope of the Changes API, 
be implemented on the adapter side, 
or potentially both?

The native indexedDB support for version tracking seems limited to unimplemented.
So I do see that as an indicator it would be useful to set up a minimal structure internally.
What are your thoughts / preferences about it?

**Comment by techfort** [comment]
I am trying to figure out the best way to do this. Loki has no hard
dependencies, only dev ones, so i'd like to keep it that way
I like the idea of something that is git-friendly in terms of api, e.g.
using commit() to "freeze" a set of changes and a push() to sync to the
server. So far i see things happening on the Changes API side, but this may
change.
Meanwhile - although this is nothing as powerful as versioning/committing -
LokiJS records have a meta object containing a revision property which
should help for diffs.
I assume @obeliskos would probably have an opinion on this too.

**Comment by obeliskos** [comment]
Sounds interesting... not sure if it would be a primary sync method but I would need to know more.  To be honest I have not needed to look into the current change api implementation.

Would you decompose the db into many files?  
What advantages would this have over other cloud storage? (A diff on the whole database might not tell me a lot).

I have thought about extending the persistenceAdapter interface to accept reference to 'this' database.  This might be used for complicated serialization scenarios where you might serialize more granulary than just 'the whole db toString()'.  Maybe we can consolidate our persistence adapters to include syncadapters and allow a wide varierty of adapters to be developed.

**Comment by seriousme** [comment]
Just tossing in my 2cts :-)

JsGit is no small module.

Using the changes api you would be able to just push the changed records from client to server.
The server can diff and send update objects back as Loki updates.(gzip will reduce the size of the updates considerably)
If you add a version indicator in the record itself  it would be fairly easy to identify update conflicts.

Alternatively the meta/rev property could be used. The update method could then be altered to check if meta/rev of the update and the record in the DB match and throw an error otherwise.

If you really want to use Git then a custom persistenceAdapter using jsGit might be a solution.
E.g. pull the changes from the server in git format, apply them to the DB data and restart the DB.

IMHO as long as there is no standard Lokijs backend, the synchronization options with backends should be optional as they could considerably impact code size and complexity.

btw: If you need easy DB synchronization then a combination of pouchDB and couchDB is way more easy ;-)
for ref: this is how couchDb does it:
http://docs.couchdb.org/en/1.6.1/api/database/bulk-api.html#db-bulk-docs

**Comment by techfort** [comment]
I agree with @obeliskos  on a persistence-adapter based implementation so - as @seriousme pointed out - we could use jsGit. This persistence adapter could be packaged as a separate lib so Loki remains without dependencies.
Having said that, i think the need of a standard LokiJS backend is gradually increasing. The meta.rev property was meant to resolve or at the very least help resolve this problem. Maybe implementing [lamport timestamp](https://github.com/techfort/LokiJS/issues/16) can come in handy too.

**Comment by PseudoCorps** [comment]
Sorry for the delayed response, I was slowly writing up something more descriptive.

I used the git example because I thought it expressed the type of functionality I was looking for in a clear way. It also offers possible solutions to other milestones.  I didn't mean to imply that it would be a dependency.(just realized #70 had closed, and I thought I read somewhere that there was interest in replication)

I agree with @obeliskos as well about opening the sync adapter(api?), it seems like it could offer a route for embedding a majority of my db setup and ingestion logic. I just have a couple additions,

#### Changes API
External Collection/s
> The track changes option for collections could be more useful if you could specify that the changes should be persisted in an external collection until they are "committed". I'm not sure if the third option is necessary, but you would be able to choose store internal(default), external, or both.
This would enable the user to make local modifications without losing them when new data is pulled. It would also enable async/on demand batch updates to a remote. 

Solution To Version History
> In the case that changes has its own collection, the add meta property would be good enough for rev and version. There would be the problem of associating the changes back to collection which could be implemented by making the collection name a property on the changes, or making the collection use collection name as its key and embedding the changes.

Features that I think would be useful design wise  for the sync adapter/api? would be:
#### Sync Adapter / API
Push 
> Enable logic to send updates to a database other than 'this' 
collect changes -> send data to listener 
optional on response hook 
then callback
On response separate from callback would be for custom cleanup or other udfs 

Pull 
> Enable logic to request changes or get a full data set from a database other than 'this' 
If is version based -> send listener version info and optional callback
else -> xhr, jsonp, etc...
after pull -> callback
Separating after pull from callback would enable custom serialization and change comparison logic

Hooks
> Before and after to enable validation or clean up.

**Comment by PseudoCorps** [comment]
In response to the question @obeliskos  asked "Would you decompose the db into many files?"
The structure that they use essentially mirrors git, so they have a tree that corresponds to a repo, and the files, data, etc are stored as nodes and edges. I found an example interface to indexeddb: https://github.com/aaronpowell/git-indexeddb/blob/master/indexeddb.js
They use hashStore to store path refs and file blobs and pathStore to store path to hash refs. (I think)
So the storage goes App/hashStore/hash -> value and App/pathStore/path -> value in indexeddb.


@seriousme I feel like its important to enable local modifications that won't be overwritten whenever you send or receive data.An example use case: From the user perspective, it's nice when an app saves the message you're typing while updating your chat (ha apparently github doesn't have this feature).
I do agree, if it looks like it will have a significant impact on the size it might not be worth it, but I imagine these are features that users would potentially be looking for and can hopefully be minimal. They could be put in a separate extra/contrib repo, which would be useful if people start contributing adapters?

**Comment by seriousme** [comment]
Hm,

Imho all doable, but it would IMHO not be Loki anymore.
Like I mentioned in a mail to @techfort  and @obeliskos about a week ago: if I wanted to do a really high perf datastore then I would build it myself and skip all the abstraction layers.

Same argument imho holds for features like git-backends, datastores-outside-this etc. 

Trying to stuff it all in the current LokiJS feels like a strech to me.

if you want to go this route then Loki could be split up in a number of modules , e.g: 
- the datastore (the current this.data array, but it could be hashmap or something totally different as well)
- the query engine (including DynamicViews)
- the indexing algorithms (unique, exact, ordered, but maybe also soundex, neuralnet etc)
- the backend sync (jsGit/JSONp/XHR)

which would mean that there would be a LokiJsToolbox and one of the databases build on top  of this toolbox would be LokiJS as we know today.

Users with less available time/skills/requirements would use the prepackaged LokiJS.
(And maybe @techfort decides to provide from the same source a LokiJS as we know now and a LokiCloudJS which automatically syncs Loki databases across the globe :-))

The experienced coders with special requirements could then build their own custom solution from the toolbox and publish it in the "examples" section of the toolbox.

My 2cts..

**Comment by techfort** [comment]
@seriousme : the usual couple of hundred dollars worth of 2c :) 

I think - and I know @seriousme has already expressed this position and @obeliskos will most likely agree - that Loki should be a minimalist, fast, user-friendly core library that allows you to do a lot by bringing in a relatively small file into your project.

I nearly feel all adapters should be moved over to a LokiAdapters repository - as it's been pointed out already - and same for all other non-core functionality.
And this is for as long as we are on v1.x.x

For v2. we could definitely go down the route of a totally "composable" db, but I would like to see that Loki has a reason to exist (i.e. has become very popular and users expect progress) before embarking on that journey.

All that said: a simple mechanism that allows to apply changes to objects and the commit them into the main collection looks like something feasible, and the basis for that already exists in the copy() feature of ResultSst, where a set of objects is basically detached from the main collection. If we implemented a commit() method that merges those changed copies into the main collection that would constitute a good enough basis for syncing / committing and it wouldn't be a dramatic change or increase in code size.

Again, I see the ability to sync local ddbs as a major step for LokiJS, and for now I'd choose the most minimalist and solid route for that, whichever one we all agree upon.
My biggest problem with this feature is that I don't have a use-case for it, so practically I'm trying to find a solution for a problem I don't have yet, but - as I said - I really appreciate the importance and relevance of it.

**Comment by seriousme** [comment]
Btw: if Loki goes down the "composable" db path then it  might be convenient to use something like http://browserify.org/ which would make it possible to generate a build for the browser while still using "require" in the modules.

**Comment by ArnoBuschmann** [comment]
Aside of Browserify one might also consider Webpack http://webpack.github.io/

**Comment by techfort** [comment]
@seriousme sometimes i think you bugged my brain with some arduino-based mind-reader thingy. @ArnoBuschmann i was not aware of webpack, looks cool, thanks for sharing!

**Comment by ArnoBuschmann** [comment]
@techfort Webpack is widespread in the Reactjs community, probably because @petehunt never stopped to point out, how good it is, haha. Oh wait, he even wrote a Webpack howto? https://github.com/petehunt/webpack-howto ;)

**Comment by seriousme** [comment]
@techfort : the cryptedFile adapter had a virtual arduino-based mind-reader hidden in there. Next patch will upgrade it from reader to writer ;-)

More seriously then:

I polished the index demo from last week a bit:   http://jsfiddle.net/jklu/46kk842d/2/
Now we got proper classes and clean separation between the index and the datastore.

It does not matter what kind of datastore you have as long as its able to do a getById these indexes can be used. (and yes I know that idIndex was the one getting this all started ;-)).
If you put your records directly in localStorage, do some git magic on them, keep them in array or in a hashmap or print them on postcards, it will work.
And id's don't need to be numerical, nor short, so you are welcome to use [UUID's](https://www.npmjs.com/package/node-uuid) or stringyfied objects (Mongo style compound indexes  anyone ? ;-)) as id's if thats what you fancy. (but don't complain about memory use then ;-))

Could this be a  starter for a /lib/indexing.js  in Loki 2.0 ? ;-)

I looked a bit at the [MongoDB docs](http://docs.mongodb.org/manual/core/query-plans/#read-operations-query-optimization) and we might take some cues from there.

The prototype shows that if you only need exact indexes, there is quite some weight that can be dropped.

**Comment by techfort** [comment]
@seriousme Superb! Glad to see some of the key-value store code put to good use (it's like a zombie in loki at the moment :) ) . Really good base for v2.0. I really like the abstraction of the index concept that you implemented, it means users may develop indexes with the most varied algorithms and get them to work, that's right down awesome. 

@obeliskos @PseudoCorps @seriousme I think we're departing quite a bit from the original RFC/enhancement. So this is my idea: 
I am happy to partially implement this commit() functionality by adopting the pattern that @PseudoCorps specified in a previous post, specifically:
## Changes API

### External Collection/s

> The track changes option for collections could be more useful if you could specify that the changes should be persisted in an external collection until they are "committed". I'm not sure if the third option is necessary, but you would be able to choose store internal(default), external, or both.
This would enable the user to make local modifications without losing them when new data is pulled. It would also enable async/on demand batch updates to a remote.

### Solution To Version History

> In the case that changes has its own collection, the add meta property would be good enough for rev and version. There would be the problem of associating the changes back to collection which could be implemented by making the collection name a property on the changes, or making the collection use collection name as its key and embedding the changes.

This should not be difficult because most of the code to create such an external collection already exists. I propose the introduction of a config property (passed in the config object into the constructor) which specifies that changes will be stored in an external collection (called something like <collectionname>-changes) which will specify <collectionname> as the collection of reference. This collection will inherit from Collection and will have a method commit() which will merge objects into the main collection (while retaining the changes). As usual suggestions/ constructive criticism is always welcome.

**Comment by seriousme** [comment]
My suggestion would be, like with the persistence adapters, to use the same interface for both internal and external change handling. That would nicely lineup for the 2.0 composable db ;-)

**Comment by PseudoCorps** [comment]
@techfort that sounds great, I was sort of worried we weren't going to come back to it. I was actually pretty surprised when I saw that the changes are flushed. The entire structure is there, so it seems doable only changing a few lines.

@seriousme, not sure what you mean by "interface for both internal and external change handling"

**Comment by techfort** [comment]
@PseudoCorps great we will go ahead with that. As for what @seriousme meant - i'm assuming by interface we're talking about the same concept that Java/C# adopt for an interface: the definition of a contract with an object (meaning the mandatory presence of certain methods with a certain signature) but  not the underlying implementation (which can change from class to class) leading do different behaviour while respecting the same contract. In this case - and I *think* that's what @seriousme meant - we could for v2.0 develop a Changes class and an ExternalChanges class which implement commit(), flush() etc. but do different things in their method. It would be up to the user to "compose" their db using this or that Changes handler. Thanks to @obeliskos  Loki possesses that kind of flexibility on storage adapters, the idea discussed here is for v2 to extend that to indexes, querying, changes, data itself and of course storage.

**Comment by seriousme** [comment]
@techfort it looks like the patch for the virtual arduino worked. We are nearly in sync. Needs one more adjustment so you *know*  instead of *think* what I meant.

More seriously then: This is exactly what I meant. With the storage adapters there used to be a difference in handling internal and external storage adapters. The externals having an specified interface contract and the internals just being called directly. By putting them on the same interface you get cleaner code and better maintainability. However if the cost of the interface (e.g. cost of converting to and from a generic function signature) is too high it will hurt performance, so the interfaces need careful design.

For this specific interface:  As it needs "commit" functionality I would suggest to group it with the transactional logic. Therefore if the user does not need transactions the overhead of commit logic can be skipped by (not) flipping an option. (commit logic typically comes at a price, e.g. the current 
collection.update uses:
``` 
try { 
   <update>  
  commit();
} 
catch (err) { rollback }
```
Even if there are no DynamicViews registered and *this.transactional == false*
which prevents the JIT from optimizing this part.

Now I'm not saying we all need to fix this in one go, but if we go for 2.0 it would be nice to refactor this.

**Comment by obeliskos** [comment]
Transactions need a re-think anyways to more properly define a transaction context across multiple operations.  My original thought was to just lift control to user and have them initiate and commit.

Could we force user to define 'change transaction' context beginnings and commits and rollbacks.

**Comment by seriousme** [comment]
Well SQL type databases like SQL lite let the user initiate a "begin transaction" 
(e.g. see https://www.sqlite.org/lang_transaction.html)
And as you say, its typically across multiple operations. (see http://en.wikipedia.org/wiki/Database_transaction)

NoSQL databases don't do SQL (hence the name ;-)) and therefore also typically don't do ACID transactions as ACID transactions in distributed systems are notoriously hard. 
(Some do, but then you lose scale out capability)

Nice set of links on the subject can be found at: http://highscalability.com/blog/2013/5/1/myth-eric-brewer-on-why-banks-are-base-not-acid-availability.html

Now the question is: which direction should Loki take: SQL + ACID or NoSQL and distributed.
Thats up for you guys to decide :-)

(btw: NoSQL also typically does not hold the notion of Join as its not a relational database,
if you want to relate docs you do that in the client application, this enables the DB to distribute data without having to worry about state)

My 1ct (2cts was too much last time ;-))

**Comment by obeliskos** [comment]
TLDR :) 

Please describe what you suggest rather than redirect an explanation.  Or don't :)

**Comment by seriousme** [comment]
Sorry for trying to educate people on the subject ;-)
My suggestion (given @techfort 's aspirations to have Loki become a distributed database)  is to stay away from transactions or anything SQL like (joins ?)  as it could bite you in the back later on.

**Comment by obeliskos** [comment]
Ah, well changes api is great but I'm not sure we need to compete with pouchdb as 'the distributed nosql database'.  Loki is great swiss army knife and to commit too deeply with other niche products diminishes loki's appeal IMO.

Yea I get too much education already without arcane relational db architecture diagrams ;)  Thanks though.  Don't take my approach to personally, techfort is also a fan of agile/scrum type development (per trello) so just trying to keep the 'stories' simple.

**Comment by PseudoCorps** [comment]
Sorry, there's a lot going on here. 
So the persisting changes on another collections is okay right? I dont think that will be a push toward a distributed architecture, but rather a simple modification that enables a lot of functionality.

**Comment by techfort** [comment]
@PseudoCorps yes , that's exactly right. It's a small enough change that opens up a few options so i'm all for that. I created an [issue](https://github.com/techfort/LokiJS/issues/125) specifically for that, and close this one which has gone well OT :) Having said that it requires a bit of design because the current changes API doesn't allow to reference a particular collection. 
@obeliskos @seriousme this has been an extremely interesting discussion. I will summarise my thoughts in a mail to you guys.
